x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -331,7 +334,8 @@ public Settings additionalSettings() {
         handlers.add(new RestXPackUsageAction());
         handlers.add(new RestReloadAnalyzersAction());
         handlers.addAll(licensing.getRestHandlers(settings, restController, clusterSettings, indexScopedSettings, settingsFilter,
-                indexNameExpressionResolver, nodesInCluster));
+            indexNameExpressionResolver, nodesInCluster));
+        handlers.add(new RestNodeEnrollmentAction());
```

Comment:
This rest handler and the corresponding transport handler are registered here in xpack core instead of security. But I wonder whether it would be better to register them under xpack security because this API assumes security being enabled (which really is by design).

When we bootstrap a new cluster, the init process manipulates the `elasticsearch.yml` file. So we have the guarantee that security is enabled. But we don't have the same guarantee for an existing cluster. IIUC, users can still explicitly disable security in v8.0. So for a cluster where security is explicitly disabled, it does not seem to have much sense to keep this API around? It's likely error out because there may not be any keystore file for it to read. It would probably be even worse if there are keystore files (maybe leftovers before security is disabled) because a successful response could give false impression of security. 

If we register this API under security, it will not be available when security is explicitly disabled which naturally solves the aforementioned issue. One advantage that I can think of for having it under core is that we can return an error response says "please enable security". But this can probably be replaced by documentation and also we don't have this type of error messages for any other security APIs.

In summary, I'd suggest to register this API and likely other related APIs, as well as the setting for enrollment mode, under security.

Reply:
I went back and forth on this decision but settled on core as these APIs felt more like cluster APIs than strictly security APIs. With that in mind it made more sense to have these in core. 

> When we bootstrap a new cluster, the init process manipulates the elasticsearch.yml file. So we have the guarantee that security is enabled. But we don't have the same guarantee for an existing cluster. IIUC, users can still explicitly disable security in v8.0. So for a cluster where security is explicitly disabled, it does not seem to have much sense to keep this API around?  

Security will default to being enabled so the init process changes to the elasticsearch.yml make no difference here. But you are right, users are still allowed to disable security if it fits their use case. We will be introducing an EnrollmentBaseRestHandler of sorts for all the Rest actions where we'd be checking if security is enabled and if enrollment mode is enabled so this solves the issue.

I don't have a strong argument in favor of keeping this in core, but I don't see a strong argument in favor of moving these to the security plugin either. Happy to discuss further though

Reply:
Sorry if this has already been a deliberated decision. I wasn't aware of the context. I am happy as long as it is an intentional change. Thanks!
> We will be introducing an EnrollmentBaseRestHandler of sorts for all the Rest actions where we'd be checking if security is enabled and if enrollment mode is enabled so this solves the issue.

If providing a more useful error message (more useful than "handler not found") is the goal, having it registered in core makes sense. I still do think this API is closer to security than core mainly by the fact that it does not work unless security is enabled. Though it's a rather weak argument for this API. What about the other API where an API key needs to be created (#72186)? Do we want to have all APIs located in the same package, i.e. core? It's possible to coordinate the API key creation in core as well, i.e rely only on the CreateApiKey Request and Response classes. So far, we do not seem to have any code creating API key outside of the security package. This might be the chance to set up an example. With our current stance towards security, I guess the ideal world is to move basic security features into `core`, which could help get rid of these discussion for good and make it easier if another plugin wants to leverage it (maybe worth to discuss separately). 

> Security will default to being enabled so the init process changes to the elasticsearch.yml make no difference here.

It maybe unlikely, but theoretically still possible that security is explicity disabled in `elasticsearch.yml` before the init process runs (maybe the user was trying to experiment before starting the init process). For the process to be foolproof, init might want to check and correct the setting if needed.

Reply:
> Sorry if this has already been a deliberated decision. I wasn't aware of the context. I am happy as long as it is an intentional change. Thanks!

No worries, always happy to revisit decisions

> If providing a more useful error message (more useful than "handler not found") is the goal, 

It's a nice effect of it, but not a hard requirement or a goal I guess. 

> It's possible to coordinate the API key creation in core as well, i.e rely only on the CreateApiKey Request and Response classes

Right! That's what we're doing so far. 

> It maybe unlikely, but theoretically still possible that security is explicity disabled in elasticsearch.yml before the init process runs (maybe the user was trying to experiment before starting the init process). For the process to be foolproof, init might want to check and correct the setting if needed.

Yes absolutely. My point was that the init process makes no difference. Whether it runs or not, the user can set the YAML file settings as they please, before or after it runs, so I agree with you that there is a chance that security is disabled and we should handle this gracefully.

Reply:
Thanks again for starting this discussion @ywangd ! I chatted with Tim this morning about this and came to the conclusion that we should move these APIs to the security plugin because; 

- In genera, moving forward we want to put things in plugins instead of core
- These APIs cannot be used when security is off ( Create enrollment token generates an API key and enroll client and enroll node validate one ) 
- We can select to have a user friendly way of notifying the users that the APIs cannot be used because security is off ( rather than returning a 404 ), and this can be a more generic approach for the rest of the security APIs too. We could register the RestHandlers even when security is off and return a helpful error message there.